data_and_models/annotations/ner/analyze.py

```diff
@@ -0,0 +1,210 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+from pathlib import Path
+from typing import Dict, List, Union
+
+import pandas as pd
+import typer
+
+"""
+This code lets analyze annotations in order to clean them for training NER models.
+
+When there are duplicated '_input_hash', the following counts might NOT be accurate:
+  - accepted texts w/ multiple labels,
+  - labels needing normalization.
+Indeed, the computation of these counts relies on '_input_hash'.
+
+This code could be used either as a script or as a function.
+
+As a script, it lets analyze an individual annotation file with a detailed report:
+```
+Usage: analyze.py [OPTIONS] INPUT_PATH
+
+Arguments:
+  INPUT_PATH    [required]
+
+Options:
+  --verbose     [default: False]
+  --help        Show this message and exit.
+```
+
+As a function, it lets analyze several annotation files with a summary table:
+```
+from pathlib import Path
+
+import pandas as pd
+from analyze import report, METRICS
+
+filenames = [
+    "annotations5_EmmanuelleLogette_2020-06-30_raw2_Disease.jsonl",
+    "annotations6_EmmanuelleLogette_2020-07-07_raw4_TaxonChebi.jsonl",
+    "annotations9_EmmanuelleLogette_2020-07-08_raw6_CelltypeProtein.jsonl",
+    "annotations14_EmmanuelleLogette_2020-09-02_raw8_CellCompartmentDrugOrgan.jsonl",
+    "annotations15_EmmanuelleLogette_2020-09-22_raw9_Pathway.jsonl",
+    "annotations10_EmmanuelleLogette_2020-08-28_raw1_raw5_10EntityTypes.jsonl",
+    "annotations12_EmmanuelleLogette_2020-08-28_raw7_10EntityTypes.jsonl",
+]
+
+indexes = []
+rows = []
+for x in filenames:
+    path = Path(x)
+    indexes.append(path.stem.split("_", maxsplit=1)[0])
+    rows.append(report(path, False))
+
+table = pd.DataFrame(rows, index=indexes, columns=METRICS)
+```
+"""
+
+METRICS = [
+    "total texts",
+    "ignored texts w/o spans",
+    "ignored texts w/ spans",
+    "accepted texts w/o spans",
+    "accepted texts w/ spans",
+    "accepted texts w/ spans (%)",
+    "duplicated '_input_hash'",
+    "accepted texts w/ multiple labels",
+    "labels needing normalization",
+]
+
+
+def report(input_path: Path, verbose: bool) -> Dict[str, Union[int, float]]:
+    print(f"Analyzing {input_path}...")
+
+    results = []
+
+    def msg(idx: int, text: str) -> None:
+        if results[idx] > 0:
+            print(text)
+
+    def info(idx: int) -> None:
+        msg(idx, f"### INFO ### This might not be expected.")
+
+    def warn(idx: int) -> None:
+        msg(idx, f"### WARNING ### There might be an issue.")
+
+    def critical(idx: int) -> None:
+        msg(idx, f"### CRITICAL ### There is an issue to investigated.")
```

Comment:
As you can see, there is a test in `msg()` (`if results[idx] > 0`).

Using a logger would then just add complexity while not adding features.

Otherwise, a logger could indeed be used.

Reply:
We decided during the planning that we keep the current implementation.

Indeed, implementing conditional logging with a logger would not be worth it at the moment.